### PR TITLE
feat: Optional Boundary Inclusiveness for Within Queries

### PR DESCRIPTION
**Pull request by cbueth**
**TLDR;** I have implemented the ability to toggle boundary inclusiveness for `within` queries across all KD-tree variants: `Mutable`, `Immutable`, `Hybrid`, and `Fixed`. 

### Description

This PR introduces the option for non-inclusive boundary searches in within queries. This change is motivated by requirements in certain information-theoretic estimators where a strict distance less than the radius is necessary (KSG: Type I vs Type 2 https://doi.org/10.1103/PhysRevE.69.066138, `distance < radius` is required for one of them). By generalizing the internal logic, we avoid code duplication while maintaining the high performance of Kiddo, especially in the leaf-level SIMD loops.

The implementation propagates an inclusive flag through the search macros for Mutable, Immutable, Hybrid, and Fixed trees. This is consistent with the findings about prunings and the changes of PR #290 & #291. New `*_with_condition` methods provide this flexibility while keeping the standard API backward compatible. So as #290 would change edge cases, with this PRs changes the user can choose to include or not to include points at exactly the maximal distance. If you want, please suggest an alternative name or API.

A full suite of tests has been added to verify boundary behavior across all tree types and distance metrics. One thing to keep in mind is whether we should eventually move towards a more structured `QueryOptions` approach if more search parameters are added in the future.

### Changes:

* I've updated the internal search macros and leaf-level distance checks to support an `inclusive: bool` flag. This flag is propagated through the query stack.
*   Introduced `*_with_condition` variants:
    *   `within` -> `within_with_condition`
    *   `within_unsorted` -> `within_unsorted_with_condition`
    *   `within_unsorted_iter` -> `within_unsorted_iter_with_condition`
    *   `nearest_n_within` -> `nearest_n_within_with_condition`
* This is fully backward compatible.

### CODE REVIEW

**Comment by sdd** [comment]
I like the idea of this. I'm leaning towards keeping the existing names for inclusive boundary operations, and suffixing with `_exclusive` otherwise. I'll have a think about this over the weekend. 

I expect that the branch would optimise out in many cases but I'd like to confirm that by looking at the asm output before accepting this as we're adding a branch in an inner loop.

**Comment by cbueth** [comment]
> I like the idea of this. I'm leaning towards keeping the existing names for inclusive boundary operations, and suffixing with `_exclusive` otherwise. I'll have a think about this over the weekend.

This is also a good choice. My first Idea was `_strict`.

> I expect that the branch would optimise out in many cases but I'd like to confirm that by looking at the asm output before accepting this as we're adding a branch in an inner loop.

Ok, let me know of that outcome.

The alternative approach I had in mind was adding `within_strict` to the macros, in the query files and in the KdTree. this would duplicate not much, but more code than now and might be more annoying to maintain, but might help the compiler optimisation. So I hope the solution with the added flags and branch will optimise out.

Have a nice weekend!

**Comment by cbueth** [comment]
I have merged `upstream/v5.x.x` and #291 and linted the files. All checks pass.
For the code coverage, are there some lines you want me to add coverage for?

**Comment by cbueth** [comment]
Would you consider these additions for `v.5.3.0` as well? If so I would rebase after merging #291.

**Comment by cbueth** [comment]
This PR has been rebased onto the merged #291.

**Comment by sdd** [comment]
Thanks for the rebase - I'll give this another read today and either merge it in, or if not, release the rest as 5.3.0 👍🏼

**Comment by sdd** [comment]
I've ran some performance benchmarks and the added conditional has not made any difference - if the compiler has not elided it completely, then the branch predictor is doing it's job and predicting it perfectly.

I think I'd still prefer if the naming scheme was `_exclusive` rather than `_with_condition` though. Once that change is made, I'll be happy to merge and get this out as part of 5.3.0 👍🏼

**Comment by cbueth** [comment]
Great news about the asm output. I have additionally added to the documentation, which was not done in any commit of this PR before.

File: src/common/generate_within_unsorted.rs [hunk 1]
```diff
@@ -17,6 +17,10 @@ macro_rules! generate_within_unsorted {
             where
                 D: DistanceMetric<A, K>,
             {
+                // Like [`within_unsorted`] but allows controlling boundary inclusiveness.
```
Comment (sdd) [inline src/common/generate_within_unsorted.rs:20]: these should be triple-/ to show in the generated docs

Comment (sdd) [reply src/common/generate_within_unsorted.rs:20]: Sorry, read them too fast before I noticed that this is not on the func def itself, ignore me :-)

**Review by sdd — APPROVED** [review]

**Comment by sdd** [comment]
Sorry for delay - merged! Will publish tonight once I get home, or tomorrow morning. Thanks!

**Comment by cbueth** [comment]
Thanks for the merge and responsive support with this and the other two PRs, greatly apprechiated!
